Add GSEA, Limma, single sample top genes

mokolodi1 opened the pull request:
In addition, I've improved the way arguments are presented for each job and also refactored the job wrapper UI code to make it easier to create new jobs.

Before this is put into production, the following migration should be run. It adds gene sets for previously run outlier analysis jobs. In addition it fixes two problems that exist with outlier analysis jobs in production. (The data set name isn't always filled in and the sample group version isn't cached per job.)
```
mongo MedBook < create_gene_sets_for_outlier_analyses.js
```

Note that this PR should be reviewed after #35 has been merged, as this branch comes from that branch.

mokolodi1 committed:
first parts of gsea tool

mokolodi1 committed:
fix spelling

mokolodi1 committed:
optimized output fields in UpDownGenes

mokolodi1 committed:
can export multiple gene set groups, add .rnk docs

mokolodi1 committed:
added gene set export/import

mokolodi1 committed:
added security docs, clarified gene sets security

mokolodi1 committed:
added comment

mokolodi1 committed:
support new job names

mokolodi1 committed:
update gene sets schema to new security model

mokolodi1 committed:
added GSEA, Limma, PairedAnalysis; update UpDownGenes

mokolodi1 committed:
remove console.log

mokolodi1 committed:
added GSEA, paired analysis, limma

mokolodi1 committed:
Merge branch 'f-gseaTool' into develop

mokolodi1 committed:
always create new outlier job, updated relevant UI message

mokolodi1 committed:
bugfix

mokolodi1 committed:
allow single sample top genes job

mokolodi1 committed:
added single sample top genes job

mokolodi1 committed:
added single sample top genes job

mokolodi1 committed:
wording fix

mokolodi1 committed:
better first-user, landing-page experience

mokolodi1 committed:
correct wording

mokolodi1 committed:
fix indentation, small refactor

mokolodi1 committed:
tell users they should contact us to try MedBook

mokolodi1 committed:
Merge branch 'f-disable-new-accounts' into develop

mokolodi1 committed:
Merge branch 'fix-docker-build' into develop

mokolodi1 committed:
changed to Dockerfile best-practices per @e-k-t's comment

mokolodi1 committed:
don't run UpdateCbioSecurity on dev machines

mokolodi1 committed:
small UI bugfixes

mokolodi1 committed:
Merge branch 'pre-gsea' into develop

mokolodi1 committed:
fix no tabs between data sets bug in exp. export

mokolodi1 committed:
add create gene sets for outlier jobs migration

This is from the old new-gsea branch from the MedBook parent repo.

mokolodi1 committed:
Merge branch 'searchable-jobs' into gsea

mokolodi1 committed:
merging in Limma/GSEA/paired jobs from patient-care

mokolodi1 committed:
don't ask for name if not logged in

mokolodi1 committed:
Merge branch 'searchable-jobs' into gsea

mokolodi1 committed:
merge in primary-collections f-gsea branch

mokolodi1 committed:
added new job-runner jobs (GSEA, Limma, etc.)

mokolodi1 committed:
Merge branch 'f-gsea' of https://github.com/UCSC-MedBook/external-tools into gsea

mokolodi1 committed:
move gene set import/export to correct location

mokolodi1 committed:
lots of small improvements and bugfixes

mokolodi1 committed:
cleaning up, refactoring #jobWrapper, consolidating files

mokolodi1 committed:
fixes issues in #37, other smallish fixes

mokolodi1 committed:
forgot to save this file before the last commit

mokolodi1 committed:
fixed sharing (adding collabs) bug

Commit:
add top 5% to outlier analysis exporter

also wasn't pulling correctly from mongo for some reason, fixed that

e-t-k commented on dev-ops/migrations/create_gene_sets_for_outlier_analyses.js:
```diff
@@ -0,0 +1,187 @@
+    gene_labels.push(dataArray[i].gene_label);
+  }
+
+  // TODO: is this _id okay? I can't figure out how to get the _id from
```
Please generate a mongo-type string _id.

`var geneSetId = (new ObjectId).valueOf()`

e-t-k commented on dev-ops/migrations/create_gene_sets_for_outlier_analyses.js:
```diff
@@ -0,0 +1,187 @@
+    bulk.insert(record);
+  }
+
+  // TODO: apparently the writeConcern needs to be an object? I tried
```
Did this get addressed or is it not necessary?

e-t-k committed:
Merge branch 'master' into outlier-exporter-script

mokolodi1 committed:
merge master into gsea

e-t-k committed:
Merge pull request #49 from UCSC-MedBook/outlier-exporter-script

Outlier exporter script

e-t-k committed:
chmod ugo+x job-runner's dockerDev.sh

e-t-k commented on patient-care/webapp/client/shared/editSampleGroup.js:
```diff
@@ -335,8 +336,10 @@ Template.formValuesFilter.onCreated(function(){
-  instance.available_filter_forms = new ReactiveVar(); 
-  instance.available_filter_forms.set([{name: "Loading forms...", formId: "placeholder_loadingforms"}]);
+  instance.available_filter_forms = new ReactiveVar();
+  instance.available_filter_forms.set([{name: "Loading forms...", urlencodedId: "placeholder_loadingforms"}]);
```
This file has a bad merge!

`urlencodedId` is an old version that was superseded by changes [here](https://github.com/UCSC-MedBook/MedBook_/pull/29/files/a099dfa926a60d728d31b77944fefe324f844ab6..16dcfe79807581c8657807b64cb00c4a87d4eb93).  

Looks like the change was reverted by [this](https://github.com/UCSC-MedBook/MedBook_/commit/f537fbaf59430a83674789d3ee408ae845d45284#diff-30b3fd25f7e664ae3f66c73044c0590c)  pull from the original patient-care.

We'll need to go through and make sure that we didn't lose any other older changes to master from that or other merges...

mokolodi1 replied:
You're right! Bizarre... For now I'm going to correct this single problem.

e-t-k replied:
Yeah, merge in all lost changes from  https://github.com/UCSC-MedBook/MedBook_/pull/29/files/a099dfa926a60d728d31b77944fefe324f844ab6..16dcfe79807581c8657807b64cb00c4a87d4eb93 first & then after that, worry about what other pulls have gotten lost.

e-t-k reviewed (changes requested):
preliminary review from skimming what files are changed; more to come. Biggest issue is that some old code has gotten merged in (see comment) and it's imperative to make sure we get rid of all of it.

mokolodi1 committed:
responded to @e-k-t's comments on migration

e-t-k commented on job-runner/webapp/server/classes/RunSingleSampleTopGenes.js:
```diff
@@ -0,0 +1,143 @@
+
+  GeneSets.rawCollection().insert({
+    // XXX: change mongo's _id creation function to use simple strings
+    _id: Random.id(),
```
`(new ObjectId).valueOf()` as for the migration

e-t-k commented on job-runner/webapp/server/classes/RunSingleSampleTopGenes.js:
```diff
@@ -0,0 +1,143 @@
+  }
+
+  GeneSets.rawCollection().insert({
+    // XXX: change mongo's _id creation function to use simple strings
```
`(new ObjectId).valueOf()` as for the migration

mokolodi1 replied:
This is actually okay as it is. I thought there might be a nice mongo option to do Meteor IDs instead of mongo ones.

mokolodi1 committed:
fix merge error located by @e-k-t

e-t-k commented on patient-care/webapp/client/shared/editSampleGroup.js:
```diff
@@ -335,8 +336,10 @@ Template.formValuesFilter.onCreated(function(){
-  instance.available_filter_forms = new ReactiveVar(); 
+  instance.available_filter_forms = new ReactiveVar();
   instance.available_filter_forms.set([{name: "Loading forms...", formId: "placeholder_loadingforms"}]);
+  instance.filter_forms_options = new ReactiveVar();
```
Please actually look at [the diff I linked](https://github.com/UCSC-MedBook/MedBook_/pull/29/files/a099dfa926a60d728d31b77944fefe324f844ab6..16dcfe79807581c8657807b64cb00c4a87d4eb93) instead of blindly changing one line.

You will see that there are more changes in that diff that have been reverted in this pull.
If you are going to fix the changes for this diff (instead of having me fix it) I need you to fix all of them.

mokolodi1 replied:
I did a spot-check for the code changes around it and all of them were carried over, which is why I didn't check every change. 

I just went through line-by-line for that diff. I found one other change that got lost. (https://github.com/UCSC-MedBook/MedBook_/commit/ce1b61f)

Should we go through the whole PR? There's 1926 files change... Or is does that diff cover all the important non-`node_modules` changes?

mokolodi1 replied:
@e-t-k and I just discussed that we'll assume this is an isolated incident until proven otherwise and chalk this up to me making an error while manually merging the file.

e-t-k reviewed (changes requested):
more changes...

mokolodi1 committed:
fix single sample top genes compare function

mokolodi1 committed:
put back change lost in merge

e-t-k committed:
Add default value type to standalone limma

Fixes 2 things:

-- right now the only possible value is gene_expression. It's annoying to have
to select gene expression every time when we know that's what the user wants.

-- the following bug @mokolodi1 mentioned in september is still active; this fix appears
to bypass it, empirically: i
"select a value type to create a limma job, then navigate away, then
go back and select a value type again. The value_type field’s
reactivity is broken and the other fields don’t show up."

Note :
this will error if there are ever 0 allowed values for what type of limma
to run.
However, if we are ever in that situation, we can't run limma anyhow, so it
does not matter.

mokolodi1 committed:
remove XXX comment: okay as is

mokolodi1 committed:
merged master into gsea

e-t-k committed:
adds (legacy) flag to Limma->GSEA [skip ci]

Now that Limma & GSEA are separate, relabel Limma->GSEA as (Legacy)
on homepage and limma-gsea page & explain its for viewing results only
Also a more friendly message than "This job has been deprecated."

Also updates GSEA to mention that it does preranked & leading edge.

e-t-k commented:
[EDIT: this has been resolved] @mokolodi1 For Limma, is it an error for the experimental sample group and the reference sample group to share samples? Right now, if they do, it fails with the cryptic message `Problem running limma`.

if it's not an error we need to allow it. If it is an error, I'd like to make the error message more explanatory.

mokolodi1 committed:
blobs.create/delete returns promise; fix: actually delete

mokolodi1 committed:
add stderr/out blobs on error in limma tool

mokolodi1 commented:
@e-t-k I've added the code to add stderr/out blobs when things go wrong with limma. It's absolutely the correct behavior for limma to fail when there are duplicate samples in the header row, but now the error message won't be so cryptic.

I also fixed a bug whereby blob files were being deleted when `Blobs.delete` was called but the mongo objects would stick around. See here: https://github.com/UCSC-MedBook/MedBook_/commit/7e76da49fe8f62aa21cee64a8f6a00b4152cd4d3#diff-94996b69ef81efb8d87b909639b75c5aR123

e-t-k committed:
Skip Leading Edge if only 1 enriched geneset found

Fixes @mokolodi1's bug via email Sept 12:
"- null pointer exception in GSEA leading edge if there’s only one
enriched gene set. Possible fix is to not run leading edge if there’s
only one enriched gene set."

If 1 geneset is found after running GSEA, skips leading edge (so no
heatmap is generated.)  Notes on the GSEA result page that this was done.

Adds key        "skipped_leading_edge" : true to the gsea job output.

e-t-k committed:
drop unnecessary variable in gsea job

e-t-k commented on tools/importers/gene_set_import.py:
```diff
@@ -0,0 +1,174 @@
+                })
+                print fieldInfo[i * 2], fieldInfo[i * 2 + 1]
+        else:
+            # TODO
```
- the fieldTypesOnly option appears to be NYI; however I don't see that it's actually ever used
- don't waste your time adding this now, but for future reference python has a csv library that can among other things create dicts using the first line as keys. see csv.dictReader

e-t-k commented on packages/blobs/server.js:
```diff
@@ -99,19 +107,27 @@ Blobs2.delete = function (selector, callback) {
+  return Q.all(rmPromises)
     .then(Meteor.bindEnvironment(function (result) {
-      callback(null, rmPromises.length);
+      var asdf = Blobs2.remove(selector, { multi: true });
```
this poorly-named variable never gets used... Also yop below, another placeholder name

mokolodi1 committed:
remove unnecessary heading on Limma/GSEA page

mokolodi1 committed:
add fields restriction to limma job publish

Previously the whole sample group options were being send down to the client, leading to very sluggish performance.

mokolodi1 committed:
update tags

e-t-k commented on patient-care/webapp/client/shared/widgets.js:
```diff
@@ -99,43 +99,135 @@ Template.editCollaborationsModal.onCreated(function() {
+});
+
+Template.editCollaborationsModal.helpers({
+  multipleObjects() {
```
Doesn't look like this is used anywhere.

e-t-k commented on patient-care/webapp/server/publications.js:
```diff
@@ -199,6 +236,158 @@ Meteor.publish("searchableJobs", function (options) {
+Meteor.publish("gseaFormData", function (maybeGeneSetId) {
+  // See below for string validation. maybeGeneSetId can be null sometimes,
+  // so Match.Optional will fail. This should be Match.Maybe but that's
+  // not available until Meteor 1.3.
```
We're using meteor 1.3 now! (I'm fine with this code as is)

e-t-k committed:
Add back python to job-runner settings

removed erroneously in 16a1d2458155ab33cbc70ae139f0fd8ca8a30f95 .
necessary for running ApplyExprAndVarianceFilters.

mokolodi1 committed:
fix @e-t-k comments, client-side permission errors for limma

mokolodi1 commented on job-runner/webapp/settings.json:
```diff
@@ -7,6 +7,7 @@
   "limma_phenotype_export":"/tools/exporters/limma_phenotype_export.py",
   "gsea_path":"/tools/gsea/rgGSEA.py",
   "gsea_jar_path":"/tools/gsea/gsea2-2.2.2.jar",
+  "python":"python",
```
I would say the better fix for this would have been to just hardcode python in the one place it's used because it’s a dictionary to itself, but that’s okay for now.

e-t-k replied:
disagree 100%. This makes it clear that python is a dependency of jobrunner & makes it very clean if we need to specify a version of python. Hardcoding one thing "just in one place" when we already have a system for it is a step down the road to madness.

e-t-k commented on job-runner/README.md:
```diff
@@ -1,5 +1,96 @@
+
+All of these files are stored as blobs with the `associated_object` being the job object. The GESA HTML folder entry-point is named `index.html`, and the heat map is named `leading_edge_heat_map_clustered.png`.
+
+### Paired Tumor Analysis (`RunPairedAnalysis`)
```
I see a bunch of code for RunPairedAnalysis, which i'm not familiar with as a feature -- but listPairedAnalysis is in .dont-load . are we not going to make it available? should drop it from the README or note why it was removed

mokolodi1 replied:
It's a feature that isn't ready for prime time just yet. Ted/Robert/Rob still have to discuss what exactly should happen to this code. For now let's keep it in the README until something is decided.

e-t-k commented on patient-care/webapp/lib/methods.js:
```diff
@@ -208,136 +208,267 @@ Meteor.methods({
+    user.ensureAccess(experimental);
+    user.ensureAccess(reference);
+
+    // TODO: make sure the groups don't share samples from the same data set
```
Fine to not do this right now, since you added the helpful error message for when limma fails due to this.
And also to be specific it's "make sure the groups don't share the same sample with expression" ; non-overlapping samples from the same data set is totally fine.

mokolodi1 committed:
undo unintentional commits

e-t-k commented on job-runner/webapp/server/classes/RunSingleSampleTopGenes.js:
```diff
@@ -0,0 +1,142 @@
+        deferred.reject(error);
+      } else {
+        // update the gene set so that people can see it
+        // NOTE:
```
Was there supposed to be a note here?

e-t-k committed:
fix no datasets warning & hide fields when needed

- for some reason, the 'ui warning message' class is set to hidden or something
on this page. I changed it to 'ui info message' and it works; not going to investigate
further at this point

- when user has no datasets, also hide " Filter top genes by percent or count? " and
"Start analysis" button  -- it is confusing to have them visisble but useless.

mokolodi1 reviewed (commented):
I haven't added "No datasets" error messages because we're hopefully going to introduce public data, so it'd be very unlikely that users would have 0 data sets.

mokolodi1 committed:
remove useless NOTE:

e-t-k commented on patient-care/webapp/client/tools/limma.html:
```diff
@@ -0,0 +1,145 @@
+    {{> afQuickField name="top_genes_count" value=1000
+        placeholder="Top N differential genes to be used in the signature"}}
+
+    {{#if duplicateSamples}}
```
This and the duplicateSamples function in limma.js are only relevant to the pairedAnalysis stuff you took out (since there is no longer an `AutoForm.getFieldValue("primary_sample_labels",
        "createLimma");` ). Will always return false (well, undefined).

mokolodi1 replied:
Good catch! Actually this was because I never implemented the thing I meant to implement. I copied it over from the paired analysis code and meant to change it to not allow job creation if there were any samples that were in both sample groups, which causes Limma to crash. I'm going to change it to a `TODO` that returns false for now.

e-t-k commented on job-runner/webapp/server/classes/RunLimma.js:
```diff
@@ -0,0 +1,201 @@
+        self.job.args.top_genes_count,
+
+        // "BH" or "none"
+        // TODO: make this an argument for the job?
```
What is BH ? Is it fine that it's hardcoded in here? would we ever want to run it with "none", whatever that means?

e-t-k commented on dev-ops/migrations/create_gene_sets_for_outlier_analyses.js:
```diff
@@ -0,0 +1,185 @@
+    var record = {
+      "Gene": oldRecord.gene_label,
+      "Sample value": oldRecord.sample_value,
+      "Background median": oldRecord.background_median,
```
Should this be omitted for top5% records? right now they all have `"Background median" : undefined,`

e-t-k reviewed (approved):
Feel free to make changes per my last 3 comments, but I'm comfortable for you to push to master & spin up the Dockers on staging at this point  while I review #39 .
